zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+        client::ChainClient,
+    },
+};
+
+// The json files are generated by invoking 'forge build' in zilliqa/src/contracts/uccb
+abigen!(
+    ValidatorManager,
+    "zilliqa/src/contracts/uccb/out/ValidatorManager.sol/ValidatorManager.json"
+);
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(value_parser = SecretKey::from_hex)]
+    secret_key: SecretKey,
+    #[clap(long, short, default_value = "config.toml")]
+    config_file: PathBuf,
+}
+
+impl zilliqa::uccb::Args for Args {
+    fn secret_key(&self) -> &SecretKey {
+        &self.secret_key
+    }
+
+    fn config_file(&self) -> &PathBuf {
+        &self.config_file
+    }
+}
+
+struct ValidatorOracle {
+    wallet: LocalWallet,
+    zq2_config: ZQ2Config,
+    validators: Vec<NodePublicKey>,
+    chain_clients: Vec<ChainClient>,
+}
+
+impl ValidatorOracle {
+    pub async fn new(secret_key: SecretKey, config: Config) -> Result<Self> {
+        let wallet = LocalWallet::from_str(secret_key.to_hex().as_str())?;
+        let mut chain_clients = vec![Self::create_zq2_chain_client(config.clone(), &wallet).await?];
+        for chain_config in config.chain_configs {
+            chain_clients.push(
+                ChainClient::new(
+                    &chain_config,
+                    H160(**config.zq2.validator_manager_address),
+                    wallet.clone(),
+                )
+                .await?,
+            );
+        }
+
+        Ok(Self {
+            wallet,
+            zq2_config: config.zq2,
+            validators: vec![],
+            chain_clients,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<()> {
+        info!(
+            "Starting validator oracle with wallet address {}",
+            self.wallet.address()
+        );
+
+        self.validators = self.get_stakers().await?;
+        info!("Validators: {}\n", serde_json::to_string(&self.validators)?);
+
+        let (sender, receiver) = watch::channel(self.validators.clone());
+
+        let mut handles = vec![];
+        for chain_client in &self.chain_clients {
+            let mut receiver = receiver.clone();
+            let chain_client = chain_client.clone();
+            let handle = tokio::spawn(async move {
+                loop {
+                    let validators = receiver.borrow_and_update().clone();
+                    if Self::update_validator_manager(&chain_client, &validators)
+                        .await
+                        .is_err()
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
```

Comment:
Did you test this? If I understand correctly, the future returned by `listen_to_staker_updates` should never return (unless there is an error), meaning the `handle.await` calls won't ever occur. The tasks will run because of `tokio::spawn` but I don't think you'll propagate errors from them like it seems you intended?

Reply:
Seems to work fine, and yes it will only really return upon an error. I suppose a loop & select! can be used instead if that's what you prefer.